tlatools/org.lamport.tlatools/src/tla2sany/semantic/BuiltInOperators.java

```diff
@@ -0,0 +1,155 @@
+package tla2sany.semantic;
+
+import util.UniqueString;
+
+/**
+ * Operators that have built-in implementations in the TLA+ interpreter.
+ * Users do not need to import modules to access these operators, and users
+ * cannot redefine these operators.
+ */
+public class BuiltInOperators {
+
+  /**
+   * Various parameters that must be set for a built-in operator.
+   */
+  public static class BuiltInOperator {
+
+    /**
+     * The operator's name, usually its string representation in source
+     * files if such a thing exists.
+     */
+    public final UniqueString Name;
+
+    /**
+     * Number of parameters the operator accepts.
+     */
+    public final int Arity;
+
+    /**
+     * The level of the operator.
+     */
+    public final int OpLevel;
+
+    /**
+     * The maximum level of the operator's parameters.
+     */
+    public final int[] ArgMaxLevels;
+
+    /**
+     * The weights of the operator's parameters.
+     */
+    public final int[] ArgWeights;
+
+    /**
+     * Constructs a new instance of the BuiltInOperator class.
+     * 
+     * @param name The operator's name.
+     * @param arity The operator's arity.
+     * @param opLevel The operator's level.
+     * @param argMaxLevels The max levels of the operator's parameters.
+     * @param argWeights The weights of the operator's parameters.
+     */
+    public BuiltInOperator(
+      String name,
+      int arity,
+      int opLevel,
+      int[] argMaxLevels,
+      int[] argWeights
+    ) {
+      this.Name = UniqueString.uniqueStringOf(name);
+      this.Arity = arity;
+      this.OpLevel = opLevel;
+      this.ArgMaxLevels = argMaxLevels;
+      this.ArgWeights = argWeights;
+    }
+  }
+  
+  /**
+   * Quick helper function to construct an array.
+   *
+   * @param vals The values to put into the array.
+   * @return An array of the given values.
+   */
+  private static int[] mk(int... vals) {
+    return vals;
+  }
+
+  /**
+   * A list of all built-in operators along with their arities and levels.
+   */
+  public static final BuiltInOperator[] Properties = {
```

Comment:
How did you generate this table? By hand? I've compared the first two columns from tlatools/org.lamport.tlatools/src/tla2sany/configuration/ConfigConstants.java#defaultConfig and they look good to me.

Reply:
I don't remember how I generated it. If you look at the sequence of commits it was part of a unit test to ensure it conforms with how the system functioned originally before it was moved to become core to the system itself.